### PR TITLE
LibCompress: Add GzipDecompressor implemenation and gunzip fontend.

### PR DESCRIPTION
**Pull request by asynts**
I am totally aware that there is another Gzip implementation in `Libraries/LibCore/Gzip.h` but I can't remove that until dynamic blocks are supported by the deflate implementation. The days of `Libraries/LibCore/puff.h` are counted too.

---

Currently, the deflate implementation doesn't support dynamic blocks which are very commonly used. Typically, very short files use fixed blocks, I've setup one such file for testing here: [simple-gzip-file.gz][1]. The following should produce "abcabc":
~~~none
pro http://static.asynts.com/08/29/simple-gzip-file.gz > simple-gzip-file.gz
gunzip -c simple-gzip-file.gz
~~~

  [1]: http://static.asynts.com/08/29/simple-gzip-file.gz

### CODE REVIEW

**Comment by alimpfard** [comment]
> curl -O http://static.asynts.com/08/29/simple-gzip-file.gz

Friendly reminder that
`pro http://static.asynts.com/08/29/simple-gzip-file.gz > simple-gzip-file.gz`
is a thing :P

**Comment by asynts** [comment]
@alimpfard Thanks! I didn't know that utility existed. I've updated the code snippet.

File: Libraries/LibCompress/Gzip.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        u8 extra_flags;
+        u8 operating_system;
+
+        bool valid_magic_number() const;
+        bool supported_by_implementation() const;
+    };
+
+    struct Flags {
+        static constexpr u8 FTEXT = 1 << 0;
+        static constexpr u8 FHCRC = 1 << 1;
```
Comment (stelar7) [inline Libraries/LibCompress/Gzip.h:62]: Something doesnt seem right about these shifts :P

File: Libraries/LibCompress/Gzip.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        size_t nread = m_block_stream.read(bytes);
+
+        if (nread < bytes.size()) {
+            m_block_stream.~DeflateDecompressor();
+            m_state = State::Idle;
+
+            LittleEndian<u32> crc32, input_size;
+            m_input_stream >> crc32 >> input_size;
+
+            // FIXME: Validate crc32 and input_size.
```
Comment (stelar7) [inline Libraries/LibCompress/Gzip.cpp:83]: There is a CRC impl in LibCrypto/Checksum that you can use :)

File: Libraries/LibCompress/Gzip.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+namespace Compress {
+
+bool GzipDecompressor::BlockHeader::valid_magic_number() const
+{
+    return identification_1 == 0x1f && identification_2 == 0x8b;
+}
+
+bool GzipDecompressor::BlockHeader::supported_by_implementation() const
+{
+    if (compression_method != 0x08) {
```
Comment (stelar7) [inline Libraries/LibCompress/Gzip.cpp:40]: Maybe a const for 0x08? (DEFLATE)

File: Libraries/LibCompress/Gzip.h (hunk 1 above)
Comment (asynts) [reply Libraries/LibCompress/Gzip.h:63]: Oops.

File: Libraries/LibCompress/Gzip.cpp (hunk 3 above)
Comment (asynts) [reply Libraries/LibCompress/Gzip.cpp:40]: There are really no other compression methods that I am aware of. The standard only defines deflate. I think the comment is enough.

File: Libraries/LibCompress/Gzip.cpp (hunk 2 above)
Comment (asynts) [reply Libraries/LibCompress/Gzip.cpp]: I've added it and it seems to work.

**Pull request merged**